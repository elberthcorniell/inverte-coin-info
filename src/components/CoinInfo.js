import { useSelector } from 'react-redux';
import { useEffect, useRef } from 'react';
import { Line, Chart } from 'react-chartjs-2';
import { formatNumber, getChartData } from '../utils';

Chart.defaults.scale.gridLines.display = false;

const CoinInfo = () => {
  const coinData = useSelector(state => state.coin);
  const desc = useRef(null);
  const {
    name,
    categories,
    description: { en: description = '' } = {},
    tickers,
    image,
    links: {
      homepage: website,
      repos_url: reposUrl,
    } = {},
    market_data: {
      market_cap: { usd: marketCap } = {},
      circulating_supply: circulatingSupply,
      total_volume: { usd: totalVolume } = {},
      current_price: { usd: price } = {},
    } = {},
  } = coinData;
  useEffect(() => {
    desc.current.innerHTML = description;
  });
  return (
    <>
      <div className="grid-1-2">
        <div className="card">
          <div style={{
            display: 'flex',
            justifyContent: 'space-between',
            alignItems: 'center',
          }}
          >
            <img src={image?.small} alt={name} className="coin-logo" />
            <h2>
              $
              {formatNumber(price, 2)}
            </h2>
          </div>
          <br />
          <div style={{ marginTop: 20 }}>
            <div className="gray-card property">
              <strong>Market Cap</strong>
              <strong>
                <span>
                  $
                  {formatNumber(marketCap, 2)}
                </span>
              </strong>
            </div>
            <div className="gray-card property">
              <strong>Circulating Supply</strong>
              <strong>
                <span>
                  $
                  {formatNumber(circulatingSupply, 2)}
                </span>
              </strong>
            </div>
            <div className="gray-card property">
              <strong>Total Volume</strong>
              <strong>
                <span>
                  $
                  {formatNumber(totalVolume, 2)}
                </span>
              </strong>
            </div>
          </div>
          <br />
          <div>
            {website?.map(link => {
              if (link.length > 0) {
                return (
                  <div key={link} className="gray-pill">
                    <a href={link}>
                      {link}
                    </a>
                  </div>
                );
              }
              return false;
            })}
            <div className="gray-pill">
              Wallet:&nbsp;
              <a href="https://inverte.io">
                Inverte
              </a>
            </div>
            {Object.keys(reposUrl)?.map(key => {
              const url = reposUrl[key][0];
              if (url) {
                return (
                  <div key={key} className="gray-pill">
                    <a href={url}>
                      {key}
                    </a>
                  </div>
                );
              }
              return false;
            })}
            {categories?.map(category => (
              <div key={category} className="gray-pill">
                {category}
              </div>
            ))}
          </div>
        </div>
        <div>
          <Line
            options={{
              responsive: true,
              legend: {
                display: false,
              },
              scales: {
                xAxes: [
                  {
                    ticks: {
                      display: false,
                    },
                  },
                ],
              },
            }}
            data={getChartData(tickers)}
          />
          <p className="gray-card" ref={desc} style={{ marginTop: 40 }} />
        </div>
      </div>
    </>
  );
};

export default CoinInfo;
