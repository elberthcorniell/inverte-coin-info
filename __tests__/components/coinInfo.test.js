import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import CoinInfo from '../../src/components/CoinInfo';
import store from '../../src/reducers';

describe('test', () => {
  const renderCoinInfo = () => {
    store.dispatch({
      type: 'SET_COIN',
      coin: {
        name: 'Bitcoin',
        symbol: 'BTC',
        links: {
          homepage: ['inverte.io'],
          repos_url: {
            Github: 'github.com',
          },
        },
        categories: ['Coin'],
      },
    });
    render(
      <Provider store={store}>
        <CoinInfo />
      </Provider>,
    );
  };

  test('Render as spected', () => {
    renderCoinInfo();
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  test('It contains a Canvas', () => {
    renderCoinInfo();
    const canvas = document.getElementsByTagName('canvas');
    expect(typeof canvas[0]).not.toBe('undefined');
  });

  test('It contains a grid', () => {
    renderCoinInfo();
    const grid = document.getElementsByClassName('grid-1-2');
    expect(typeof grid[0]).not.toBe('undefined');
  });
});
