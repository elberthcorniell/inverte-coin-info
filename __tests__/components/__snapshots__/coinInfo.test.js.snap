// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Render as spected 1`] = `"<div><div class=\\"grid-1-2\\"><div class=\\"card\\"><div style=\\"display: flex; justify-content: space-between; align-items: center;\\"><img alt=\\"Bitcoin\\" class=\\"coin-logo\\"><h2>$0.00</h2></div><br><div style=\\"margin-top: 20px;\\"><div class=\\"gray-card property\\"><strong>Market Cap</strong><strong><span>$0.00</span></strong></div><div class=\\"gray-card property\\"><strong>Circulating Supply</strong><strong><span>$0.00</span></strong></div><div class=\\"gray-card property\\"><strong>Total Volume</strong><strong><span>$0.00</span></strong></div></div><br><div><div class=\\"gray-pill\\"><a href=\\"inverte.io\\">inverte.io</a></div><div class=\\"gray-pill\\">Wallet:&nbsp;<a href=\\"https://inverte.io\\">Inverte</a></div><div class=\\"gray-pill\\"><a href=\\"g\\">Github</a></div><div class=\\"gray-pill\\">Coin</div></div></div><div><canvas height=\\"0\\" width=\\"0\\" style=\\"display: block;\\" class=\\"chartjs-render-monitor\\"></canvas><p class=\\"gray-card\\" style=\\"margin-top: 40px;\\"></p></div></div></div>"`;
